src/shared/install.c

```diff
@@ -2979,22 +3045,49 @@ static int preset_prepare_one(
                 return 0;
         }
 
-        r = query_presets(name, presets);
+        r = query_presets(name, presets, &instance_name_list);
         if (r < 0)
                 return r;
 
         if (r > 0) {
-                r = install_info_discover(scope, plus, paths, name, SEARCH_LOAD|SEARCH_FOLLOW_CONFIG_SYMLINKS,
-                                          &i, changes, n_changes);
-                if (r < 0)
-                        return r;
+                if (instance_name_list) {
+                        char **s;
+                        STRV_FOREACH(s, instance_name_list) {
+                                r = install_info_discover(scope, plus, paths, *s, SEARCH_LOAD|SEARCH_FOLLOW_CONFIG_SYMLINKS,
+                                                          &i, changes, n_changes);
+                                if (r < 0)
+                                        return r;
 
-                r = install_info_may_process(i, paths, changes, n_changes);
-                if (r < 0)
-                        return r;
-        } else
-                r = install_info_discover(scope, minus, paths, name, SEARCH_FOLLOW_CONFIG_SYMLINKS,
-                                          &i, changes, n_changes);
+                                r = install_info_may_process(i, paths, changes, n_changes);
+                                if (r < 0)
```

Comment:
Maybe add a short helper func that would combine `install_info_discover` and `install_info_may_process` and use that in both branches here.

Reply:
Thanks, I was considering adding that, but I saw the similar pattern (`install_info_discover` followed by `install_info_may_process`) showed up multiple times in the file. 

Do I create a helper function replacing all of the similar patterns, or would it be sufficient to replace the current ones?

Reply:
Preferably all. If there's more than one or two, please make this a separate patch.

Reply:
Sure, I will be drafting another patch shortly :).

Reply:
Hi, I attempted to create a PR in https://github.com/systemd/systemd/pull/9949, let me know if it is done correctly when you have time :), thanks.